### ISSUE TITLE
[x86/Linux] Enable FEATURE_FIXED_OUT_ARGS

### ISSUE THREAD

**Issue by seanshpark** [dotnet/coreclr#10012]
I'd like to add this issue for `FEATURE_FIXED_OUT_ARGS` enabling on x86 Linux.
Currently there exist several stack alignment issues as ESP is changed while making a call.

With discussion with @parjong, we can do as these steps;
1. Use `FEATURE_FIXED_OUT_ARGS` directive to split or insert to current code
2. Turn on `FEATURE_FIXED_OUT_ARGS` locally for development and testing
3. Send PRs one by one and maybe add a reference to this issue.
4. If required, introduce a new definition in cmake or other place to sync GC/VM/JIT
5. When all things goes ok, turn on this directive
6. Optionally, remove(clean) stack alignment codes, current code with `UNIX_X86_ABI`, as it may have failures in unit test and may confuse developers.

**Comment by janvorli** [comment]
@seanshpark thank you for sharing your plan.
As for #3, I would prefer to have one PR that would contain all the logical changes, each change being a separate commit. It would be very difficult to keep track of all the changes that were made if they were separate PRs and it would make my reasoning about them during code reviewing very difficult.
As for #4, I am not sure what that means, can you please give me some details?

**Comment by parjong** [comment]
@janvorli As this is a kind of ABI (contract between VM and JIT), we need to modify both VM and JIT all at once. 

For example, unwinder need to be changed according to ``FEATURE_FIXED_OUT_ARGS``, but ``FEATURE_FIXED_OUT_ARGS`` is defined only in JIT, and thus we currently have to use some indirect approach (such as using _TARGET_X86_).

Item 4 is a plan to introduce a global macro that affects both VM and JIT.

**Comment by BruceForstall** [comment]
cc @BruceForstall @dotnet/jit-contrib

**Comment by parjong** [comment]
\CC @parjong

**Comment by BruceForstall** [comment]
@parjong @seanshpark Can you give me some more context as to *why* you want to enable this for x86?

I can see a couple problems:
1. I don't know if the x86 GC info can handle the outgoing arg space. I think it is built around doing "push" to prepare outgoing arguments. But this requires more investigation.
2. The outgoing arg space will have to be filled by "mov [esp+x], reg" instructions, which are bigger than "push" instructions, so code overall will be bigger. Also, IIRC, you can "push [mem]" but you can't "mov [esp+x], [mem]". Not sure if we generate that pattern, though.

**Comment by parjong** [comment]
@BruceForstall It seems that ``FEATURE_FIXED_OUT_ARGS `` is required to guarantee  16-byte staack alignment.

As I know, Execution can be interrupted at any point in the function/funclet body, which means that we encounter an unaligned stack frame if execution is interrupted while passing arguments to callee via push. From that point, we internally conclude that stack frame should be stable inside function/funclect body in order to avoid this kind of issues, and decide to enable ``FEATURE_FIXED_OUT_ARGS``.

Could you let me know if there is something wrong?

**Comment by BruceForstall** [comment]
Your logic makes sense. I guess the question is: who is checking the alignment, and when?

E.g., does the stack need to always be aligned when single stepping in the debugger? That obviously can't be true when single stepping through a "call" instruction, which pushes the return address.

So if we "push" arguments for a call, who will interrupt and check? The GC?

**Comment by parjong** [comment]
@BruceForstall There is no explicit check although I am currently working on an alignment check for debug build. 

The main issue is that GCC/Clang selects instructions (such as ``movaps``) which has alignment requirement. under the assumption that stack is 16-byte aligned. When this alignment is broken, it raises segmentation fault which we currently attempt to address.

In other words, 16-byte alignment should be satisfied at the managed-unmanaged boundary.

For the last question, I guesses that GC may be, and EH may be for rude abort scenario (not sure).

**Comment by BruceForstall** [comment]
It seems like if we fix #10122, that would be sufficient for our purposes (16-byte aligned stacks at call sites), and we would probably get better code gen.

It seems like anyone interrupting execution at an arbitrary point in the function already needs to handle possible unaligned stack, due to the "call" instruction. E.g., it's not like GC interruption will then immediately start executing on the same stack it just interrupted, and if it did, it would need to dynamically align the stack anyway.

FWIW, without doing any measurements myself, I was told that `movups` and `movaps` have equivalent performance on modern processors anyway, but I guess that doesn't matter since gcc/clang already use them...

**Comment by parjong** [comment]
As I know, prologue/epilogue are marked as non-interruptible region, which means that VM cannot interrupt execution just after "call" execution. VM should wait till the end of prologue or hijack that thread.

If VM attempts to interrupt execution just before "call" is executed, it should be fine as stack should be aligned before call is made.

This may lead to some implicit assumption that stack is aligned when execution is interrupted (although I'm not sure about that).

I definitely agree with you that #10122 will make it better, but I'm not sure whether it would be sufficient.

FYI, [this change](https://github.com/parjong/coreclr/commit/565d3c22ad3fb68dd47f9a31e683a28143d58465) seems to be effective in finding stack alignment issues.

**Comment by janvorli** [comment]
@BruceForstall I was thinking about it a bit and I got an idea that would get us the benefit of both worlds - FEATURE_FIXED_OUT_ARGS that ensures that the caller SP is the same for all callees in a method and using push for storing arguments. If we added size of stack arguments to esp and then pushed the arguments, we would get that. The runtime cost would be just the added instruction to to augment esp. In the current way, to ensure alignment, we need to move esp in most cases anyways, so there should be no perf loss. 
Can you see any pitfalls in that?

**Comment by BruceForstall** [comment]
We would also need to sub esp after the call to restore the fixed outgoing arg size space, since in the current x86 convention non-varargs functions are callee-pop.

Is the requirement (or desire) to have all callees of a method have the same caller-SP due to the WIN64EXCEPTIONS EH code that we're using? Since all current targets that use that have fixed outgoing args ABIs?

**Comment by parjong** [comment]
It seems that it is non-trivial to implement correct unwinder without FEATURE_FIXED_OUT_ARGS. Is that related with this?

**Comment by janvorli** [comment]
@BruceForstall it is not a requirement, but there were couple of places where we had or still will have to do some fixups due to the fact that the caller SP is variable. It seems that with FEATURE_FIXED_OUT_ARGS, things would get cleaner.
Here is a list of several related things, I may have missed some.
* Managed and native unwinding doesn't match w.r.t. the unwound esp. Native (libunwind) unwinder gets the caller esp while the managed unwinder gets the esp before the caller pushed the stack arguments. So when we are unwinding stack during exception handling and unwind from native to managed frame or from managed to native frame, the esp we get is incorrect. We had to make jitter to always generate ebp frames to semi-compensate for it (it still gets a wrong esp, but the next unwind fixes it). 
With FEATURE_FIXED_OUT_ARGS, the esp would match. Without it, we would need to keep a table for each JIT helper function that would allow us to lookup stack arguments size for a jit helper when we unwind from native to managed frame and augment the esp accordingly.
It is possible though that we could be able to modify the unwinder and the GC info generated by the JIT to let it work with the caller esp like the native unwinder does.
* Managed unwinder had to be modified to take into account the explicit stack alignment that is inserted to  make the caller ESP aligned to 16 bytes.
* CONTEXT had to be modified to keep both caller esp (named ResumeESP) and unwound esp (the esp before the stack arguments were pushed), because the caller esp is needed for restoring context during EH and unwound esp is needed for the managed unwinder and other purposes.

**Comment by seanshpark** [comment]
@BruceForstall , yes, I'm working on this and WIP PR is #10034. Argument is setup with "mov [esp+n], REG".

Current status is `helloworld` works but there are about 800 failures with priority 1 unit test (about 11,000 test cases?), so didn't sent any review request. #10333 and #10373 are two items while I was working and now it is OK. I'm thinking to close them when it's merged. 
Now I'm working on to find why I getting an assert
```
Assert failure(PID 4587 [0x000011eb], Thread: 4587 [0x11eb]): !intersect( ptrArgs, argMask)
    File: /home/maxwell/netcore/coreclr/src/vm/eetwain.cpp Line: 2664
```

> After calls, we'll need a sub esp, stack-arg-size unless we go to cdecl

Yes, current patch does this and when CDECL is merged, I'll rebase and change the related codes.

> Also, have you investigated whether the x86 GC info can handle either of these modes?

Not not yet but  I'd be glad to know about this.  I've not much of experience with GC codes. I had some lines commented out and added a `TODO-FOA` where something like a GC related. 

> It will be interesting to see what the code size loss will be due to call un-nesting

I'm not yet on call un-nesting but I got some larger instructions cause of `struct` to temp copy variables, that need to be fixed when the whole thing works. Maybe I need to implement nested call things to fix above assert and unit test faillures that I didn't yet investigate.

**Comment by parjong** [comment]
\CC @wateret

**Comment by wateret** [comment]
@BruceForstall 

> Also, have you investigated whether the x86 GC info can handle either of these modes?

@janvorli and I had similar discussion. After I did some investigation on that, [here](https://github.com/dotnet/coreclr/issues/10599#issuecomment-293516071) is my conclusion for that (which is basically no easy way to support for fixed out args). It would be glad if you have any opinion on that.

**Comment by seanshpark** [comment]
@janvoli, could we remove Milestone 2.0.0 ? I don't think x86 is prepared for that and FOA also.
